core/src/main/java/org/apache/calcite/plan/RelOptRule.java

```diff
@@ -100,7 +100,7 @@ public RelOptRule(RelOptRuleOperand operand,
     if (description == null) {
       description = guessDescription(getClass().getName());
     }
-    if (!description.matches("[A-Za-z][-A-Za-z0-9_.():]*")) {
+    if (!description.matches("[A-Za-z][-A-Za-z0-9_.(),\\[\\]:]*")) {
       throw new RuntimeException("Rule description '" + description
```

Comment:
Why add `[` and `]` ?

Reply:
Add [ and ] because the string representation of RelCollation looks like "[]", "[2]", "[0 DESC, 1]", "[0 DESC, 1 ASC NULLS LAST]"

Reply:
Broadly seems good for me, just one question: for what reference do we make such a pattern, cause you have changed it, i want to know the source(reason) the description "should" look like what you have proposed.

Reply:
In fact I'm not sure the exact purpose of the pattern, maybe for readability. What I have done is just that allow the basic RelTrait to be part of the rule description because Calcite use it in rule description in some case.